scripts/Interface/ui/sans_isis/run_selector_widget.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import (absolute_import, division, print_function)
+
+from PyQt4 import QtGui, QtCore
+
+import ui_run_selector_widget
+from PyQt4.QtCore import pyqtSignal
+from mantidqtpython import MantidQt
+
+
+class RunSelectorWidget(QtGui.QWidget, ui_run_selector_widget.Ui_RunSelectorWidget):
+    manageDirectories = pyqtSignal()
+    browse = pyqtSignal()
+    addRuns = pyqtSignal()
+    removeRuns = pyqtSignal()
+    removeAllRuns = pyqtSignal()
+
+    def __init__(self, parent=None):
+        super(RunSelectorWidget, self).__init__(parent)
+        self.setupUi(self)
+        self._connect_signals()
+
+    def setupUi(self, other):
+        ui_run_selector_widget.Ui_RunSelectorWidget.setupUi(self, other)
+        self.runList.setSelectionMode(QtGui.QAbstractItemView.ExtendedSelection)
+
+    def show_file_picker(self, extensions, search_directories):
+        previous_directories = self._previous_directory_settings()
+        default_directory = search_directories[0]
+        directory = self._previous_or_default_directory(previous_directories, default_directory)
+        file_filter = self._filter_for_extensions(extensions)
+        chosen_files = QtGui.QFileDialog.getOpenFileNames(self, "Select files", directory, file_filter)
+        if chosen_files:
+            self._store_previous_directory(previous_directories, chosen_files[0])
+        return chosen_files
+
+    def _previous_directory_settings(self):
+        previous_directories = QtCore.QSettings()
+        previous_directories.beginGroup("CustomInterfaces/SANSRunWindow/AddRuns")
+        return previous_directories
+
+    def _previous_or_default_directory(self, settings, default):
+        return settings.value("InPath", default)
+
+    def _store_previous_directory(self, settings, path):
+        previous_file = QtCore.QFileInfo(path)
+        settings.setValue("InPath", previous_file.absoluteDir().absolutePath())
+
+    def _filter_for_extensions(self, extensions):
+        if extensions:
```

Comment:
I think you should have the if in the presenter. Then you can test if the extension is add iff there is an extension to add

Reply:
The format of the filter string is also a Qt specific detail.

Reply:
could you add qt to the name, so that its clear? If its a QString I think python handles conversion to a string internally.

Reply:
I wasn't referring to the encoding (although some of that has to be done manually). This is in the view and untested in the same way that we don't test the content of the error/warning messages or the layout of the view, it's view implementation specific and tested with our eyes (although I am now wondering if I can remove the `if` and just have `"Files ( *" + " *".join(extensions) + ")"`).

To me at least, MVP doesn't mean take all of the logic out of the view, simply take all view independent logic out of the view. To completely and cleanly eliminate logic from the view one needs an extension of MVP architecture with and extra layer - somewhere to put view dependent logic outside of the view itself, something more akin to some variants of MVVM/[Clean Architecture](https://8thlight.com/blog/uncle-bob/2012/08/13/the-clean-architecture.html).

Putting view implementation dependent logic into the presenter means you mix application specific and view dependent logic which, among other things, makes replacing/changing the view more difficult.